Stop using WTF_ALLOW_UNSAFE_BUFFER_USAGE in WTF::Deque

cdumez opened the pull request:
#### 12365d11889effd1b304f1bc6e795c6a9abf0937
<pre>
Stop using WTF_ALLOW_UNSAFE_BUFFER_USAGE in WTF::Deque
<a href="https://bugs.webkit.org/show_bug.cgi?id=283493">https://bugs.webkit.org/show_bug.cgi?id=283493</a>

Reviewed by Darin Adler.

* Source/WTF/wtf/Deque.h:
(WTF::inlineCapacity&gt;::Deque):
(WTF::inlineCapacity&gt;::destroyAll):
(WTF::inlineCapacity&gt;::expandCapacity):
(WTF::inlineCapacity&gt;::append):
(WTF::inlineCapacity&gt;::prepend):
(WTF::inlineCapacity&gt;::removeFirst):
(WTF::inlineCapacity&gt;::removeLast):
(WTF::inlineCapacity&gt;::remove):
(WTF::inlineCapacity&gt;::after const):
(WTF::inlineCapacity&gt;::before const):
* Source/WTF/wtf/Vector.h:
(WTF::VectorBufferBase::span):
(WTF::VectorBufferBase::span const):

Canonical link: <a href="https://commits.webkit.org/287067@main">https://commits.webkit.org/287067@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d977a1aba9c2b7e825dbb931174c26b895bddd40

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/78020 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/57055 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/31395 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/82670 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/29276 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/66215 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/5348 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/61093 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/19017 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/81088 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/51088 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/67444 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/41400 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/48449 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/24562 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/27621 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/71163 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/69545 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/24910 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/84032 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/77255 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/5387 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/3638 "Found 1 new test failure: ipc/create-media-source-with-invalid-constraints-crash.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/69314 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/5543 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/66950 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/68568 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/12569 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/10759 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/99568 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12093 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/5335 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/21753 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/5325 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/8757 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/7112 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WTF/wtf/Deque.h:
```diff
@@ -79,12 +77,12 @@ template<typename T, size_t inlineCapacity> class Deque final {
 
-    T& last() { RELEASE_ASSERT(m_start != m_end); return *(--end()); }
-    const T& last() const { RELEASE_ASSERT(m_start != m_end); return *(--end()); }
+    T& last() { return m_end ? m_buffer.capacitySpan()[m_end - 1] : m_buffer.capacitySpan().back(); }
+    const T& last() const { return m_end ? m_buffer.capacitySpan()[m_end - 1] : m_buffer.capacitySpan().back(); }
```
I don't understand how the call to `back()` is useful. Seems that if `m_end` is 0 then this needs to crash or assert. Letting `m_end - 1` underflow would already do that and `back()` might assert or might return an incorrect result.

cdumez replied:
It's just the way Deque works. It uses a VectorBuffer as a ring buffer. It is entirely possible for m_end to be *before* m_start and for it to be 0. When m_end is 0, the last item is at the end of the VectorBuffer. 

If it helps, the code used to do: `*(--end())` and `end()` would return `iterator(this, m_end)`.

the `operator--` for the Deque iterator is:
```
Iterator& operator--() { Base::decrement(); return *this; }
```
And Base::decrement is:
```
template<typename T, size_t inlineCapacity>
inline void DequeIteratorBase<T, inlineCapacity>::decrement()
{
    checkValidity();
    ASSERT(m_index != m_deque->m_start);
    ASSERT(m_deque->m_buffer.capacity());
    if (!m_index)
        m_index = m_deque->m_buffer.capacity() - 1;
    else
        --m_index;
    checkValidity();
}
```

Notice that decrement() handles `m_index` being 0 and falls back to the end of the capacity buffer.

darinadler commented on Source/WTF/wtf/Vector.h:
```diff
@@ -276,16 +281,31 @@ struct VectorTypeOperations
 
+    static void move(std::span<T> src, std::span<T> dst)
+    {
+        VectorMover<VectorTraits<T>::canMoveWithMemcpy, T>::move(std::to_address(src.begin()), std::to_address(src.end()), std::to_address(dst.begin()));
```
This doesn't check the size of `dst`.

cdumez replied:
Yes, Ideally it would but this is not a regression in this PR. I still think this is a step in the right direction. I'm trying to reduce the risk of introducing a perf regression in this PR by keeping the change minimal.

That said, A/B testing has been down for a few days so I haven't been able to do any perf testing for this patch yet.

darinadler commented on Source/WTF/wtf/Vector.h:
```diff
@@ -276,16 +281,31 @@ struct VectorTypeOperations
 
+    static void moveOverlapping(std::span<T> src, std::span<T> dst)
+    {
+        VectorMover<VectorTraits<T>::canMoveWithMemcpy, T>::moveOverlapping(std::to_address(src.begin()), std::to_address(src.end()), std::to_address(dst.begin()));
```
This doesn't check the size of `dst`.

darinadler commented on Source/WTF/wtf/Vector.h:
```diff
@@ -276,16 +281,31 @@ struct VectorTypeOperations
 
+    static void uninitializedCopy(std::span<const T> src, std::span<T> dst)
+    {
+        VectorCopier<VectorTraits<T>::canCopyWithMemcpy, T>::uninitializedCopy(std::to_address(src.begin()), std::to_address(src.end()), std::to_address(dst.begin()));
```
This doesn't check the size of `dst`.

cdumez committed:
Stop using WTF_ALLOW_UNSAFE_BUFFER_USAGE in WTF::Deque
https://bugs.webkit.org/show_bug.cgi?id=283493

Reviewed by Darin Adler.

* Source/WTF/wtf/Deque.h:
(WTF::inlineCapacity>::Deque):
(WTF::inlineCapacity>::destroyAll):
(WTF::inlineCapacity>::expandCapacity):
(WTF::inlineCapacity>::append):
(WTF::inlineCapacity>::prepend):
(WTF::inlineCapacity>::removeFirst):
(WTF::inlineCapacity>::removeLast):
(WTF::inlineCapacity>::remove):
(WTF::inlineCapacity>::after const):
(WTF::inlineCapacity>::before const):
* Source/WTF/wtf/Vector.h:
(WTF::VectorBufferBase::span):
(WTF::VectorBufferBase::span const):

Canonical link: https://commits.webkit.org/287067@main
